# My (uwsd) works on:

> ## Data Analysis with Python: Zero to Pandas
By **[Aakash](https://twitter.com/aakashns)**, August 2020, [Jovian ML certified Course](https://jovian.ml/learn/data-analysis-with-python-zero-to-pandas)
> ### Week 1 - Part 1

[![Python](https://img.shields.io/badge/python-3.8.5-blue)]()
[![pandas](https://img.shields.io/badge/pandas-1.1.0-blue)]()
[![numpy](https://img.shields.io/badge/numpy-1.19.1-blue)]()
[![matplotlib](https://img.shields.io/badge/matplotlib-3.3.0-blue)]()
[![seaborn](https://img.shields.io/badge/seaborn-0.10.1-blue)]()
<!-- [![Tensorflow](https://img.shields.io/badge/tensorflow-2.4.0--dev20200801-blue)]() -->

---

In [1]:
%reload_ext watermark
%watermark -a uwsd -u -v -n -t -p numpy,pandas,matplotlib,seaborn,jovian

<IPython.core.display.Javascript object>

uwsd 
last updated: Sun Aug 16 2020 15:44:29 

CPython 3.8.5
IPython 7.17.0

numpy 1.19.1
pandas 1.1.0
matplotlib 3.3.1
seaborn 0.10.1
jovian 0.2.18


In [2]:
import numpy as np
import pandas as pd

%matplotlib inline
from matplotlib import pyplot as plt
%matplotlib inline
plt.rcParams.update({
    'figure.figsize': (12, 8),
    'axes.spines.right': False,
    'axes.spines.left': True,
    'axes.spines.top': False,
    'axes.spines.bottom': True})

import seaborn as sns

import jovian

---

# Week 1: First Steps with Python and Jupyter 
## Part 1 of "A Gentle Introduction to Programming with Python"

The following topics are covered in this tutorial:

* Performing arithmetic operations using Python
* Solving multi-step problems using variables
* Evaluating conditions using Python
* Combining conditions with logical operators
* Adding text styles using Markdown

---
## Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter. 

> **Working with Cells**: To create a new cell within Jupyter, you can select "Insert > Insert Cell Below" from the menu bar or just press the "+" button on the toolbar. You can also use the keyboard shortcut `Esc+B` to create a new cell. Once a cell is created, click on it to select it. You can then change the cell type to code or markdown (text) using "Cell > Cell Type" menu option. You can also use the keyboard shortcuts `Esc+Y` and `Esc+M`. Double click a cell to edit the content within the cell. To apply your changes and run a cell, use the "Cell > Run Cells" menu option or click the "Run" button on the toolbar or just use the keyboard shortcut `Shift+Enter`. You can see a full list of keyboard shortcuts using the "Help > Keyboard Shortcuts" menu option.

Run the code cells below to perform calculations and view their result. Try changing the numbers and run the changed cells again to see updated results. Can you guess what the `//`, `%` and `**` operators are used for?

---

In [3]:
2 + 3 + 9

14

In [4]:
99 - 73

26

In [5]:
23.54 * -1432

-33709.28

In [6]:
100 / 7

14.285714285714286

In [7]:
100 // 7

14

In [8]:
100 % 7

2

In [9]:
5 ** 3

125

As you might expect, certain operators like `/` and `*` take precendence over other operators like `+` and `-` as per mathematical conventions. You can use parantheses i.e. `(` and `)` to specify the order in which operations are performed.

In [10]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

---
Python supports the following arithmetic operators:

| Operator   | Purpose           | Example     | Result    |
|------------|-------------------|-------------|-----------|
| `+`        | Addition          | `2 + 3`     | `5`       |
| `-`        | Subtraction       | `3 - 2`     | `1`       |
| `*`        | Multiplication    | `8 * 12`    | `96`      |
| `/`        | Division          | `100 / 7`   | `14.28..` |
| `//`       | Floor Division    | `100 // 7`  | `14`      |    
| `%`        | Modulus/Remainder | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |

---

## Solving multi-step problems using variables

Let's try solving the following word problem using Python: 

> A grocery store sells a bag of ice for $1.25, and makes 20% profit. If it sells 500 bags of ice, how much total profit does it make?

We can list out the information provided, and gradually convert the word problem into a mathematical expression which can be evaluated using Python. 

*Cost of ice bag ($)* = 1.25

*Profit margin* = 20% = .2

*Profit per bag ($)* = profit margin * cost of ice bag = .2 * 1.25

*No. of bags* = 500

*Total profit* = no. of bags * profit per bag = 500 * (.2 * 1.25)

---

In [11]:
500 * (.2 * 1.25)

125.0

---
Thus, the grocery store makes a total profit of $125.0 . While this is a reasonable way to solve a problem, it's not quite clear by looking at the code cell what the numbers represent. We can give names to each of the numbers by creating Python *variables*.

> **Variables**: While working with a programming language such as Python, informations is stored in *variables*. You can think of variables as containers for storing data. The data stored within a variable is called it's *value*.

In [12]:
cost_of_ice_bag = 1.25

In [13]:
profit_margin = .2

In [14]:
number_of_bags = 500

The variables `cost_of_ice_bag`, `profit_margin` and `number_of_bags` now contain the information provided in the word problem. We can check the value of a variable by typing its name into a cell, and we can combine variables using arithmetic operations to create other variables.

> Tip: While typing the name of an existing variable in a code cell within Jupyter, you can type the first few characters and press the `Tab` key to autocomplete the variable's name. Try typing `pro` in a code cell below and press `Tab` to autocomplete to `profit_margin`.

In [15]:
profit_margin

0.2

In [16]:
profit_per_bag = cost_of_ice_bag * profit_margin

In [17]:
profit_per_bag

0.25

In [18]:
total_profit = number_of_bags * profit_per_bag

In [19]:
total_profit

125.0

---
If you try to see the value of a variable that has not been *defined* i.e. given a value using the assignment statement `variable_name = value`, then Python shows an error.

---

In [22]:
try:
    net_profit
except NameError as ne:
    print(f'Error: {ne}')

Error: name 'net_profit' is not defined


---
Storing and manipulating data using appropriately named variables is a great way to explain what your code does.

Let's display the result of the word problem using a friendly message. We can do this using the `print` *function*.

> **Functions**: A function is a reusable set of instructions. A function takes one or more inputs, performs certain operations, and often returns an output. Python provides many in-built functions like `print`, and also allows us to define our own functions.

In [23]:
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


> **`print`**: The `print` function is used to display information. It takes one or more inputs which can be text (within quotes e.g. `"this is some text"`), numbers, variables, mathematical expressions etc. We'll learn more about variables & functions in the next section.

Creating a code cell for each variable or mathematical operation can get tedious. Fortunately, Jupyter allows you write multiple lines of code within a single code cell. Let's rewrite the solution to our word problem within a single cell.

In [24]:
# Store input data in variables
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


---
Note that we're using the `#` character to add *comments* within our code. 

> **Comments**: Comments and blank lines are ignored during execution, but they are useful for providing information to other humans (including yourself) about what the code does. Comments can be inline (at the end of some code), on a separate line, or even span multiple lines. 

Inline and single line comments start with `#`, whereas multi-line comments begin and end with three quotes i.e `"""`. Here are some examples of code comments:

---

In [25]:
my_favorite_number = 1 # an inline comment

In [26]:
# This comment gets its own line
my_least_favorite_number = 3

In [28]:
"""This is a multi-line comment.
Write as little or as much as you'd like.

Comments are really helpful for people reading
your code, but try to keep them short & to-the-point.

Also, if you use good variable names, then your code is
often self explanatory, and you may not even need comments!
"""
a_neutral_number = 5

---
## Evaluating conditions using Python

Apart from arithmetic operations, Python also provides several oprations for comparing numbers & variables.

| Operator    | Description                                                     |
|-------------|-----------------------------------------------------------------|
| `==`        | Check if operands are equal                                     |
| `!=`        | Check if operands are not equal                                 |
| `>`         | Check if left operand is greater than right operand             |
| `<`         | Check if left operand is less than right operand                |
| `>=`        | Check if left operand is greater than or equal to right operand |
| `<=`        | Check if left operand is less than or equal to right operand    |

The result of a comparision operation is either `True` or `False` (note the uppercase `T` and `F`). These are special keywords in Python. Let's try out some experiment with comparision operators.

---

In [29]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [30]:
# Equality check - True
my_favorite_number == 1

True

In [31]:
# Equality check - False
my_favorite_number == my_least_favorite_number

False

In [32]:
# Not equal check - True
my_favorite_number != a_neutral_number

True

In [33]:
# Not equal check - False
a_neutral_number != 3

False

In [34]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

True

In [35]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

False

In [36]:
# Less than check - True
my_favorite_number < 10

True

In [37]:
# Less than check - False
my_least_favorite_number < my_favorite_number

False

In [38]:
# Greater than or equal check - True
my_favorite_number >= 1

True

In [39]:
# Greater than or equal check - False
my_favorite_number >= 3

False

In [40]:
# Less than or equal check - True
3 + 6 <= 9

True

In [41]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3

False

---
Just like arithmetic operations, the result of a comparison opration can also be stored in a variable.

---

In [42]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10
print("Is the ice bag expensive?", is_ice_bag_expensive)

Is the ice bag expensive? False


---
## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evalute to `True`. Otherwise it returns `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |


In [43]:
my_favorite_number

1

In [44]:
my_favorite_number > 0 and my_favorite_number <= 3

True

In [45]:
my_favorite_number < 0 and my_favorite_number <= 3

False

In [46]:
my_favorite_number > 0 and my_favorite_number >= 3

False

In [47]:
True and False

False

In [48]:
True and True

True

---
The `or` operator returns `True` if at least one of the conditions evalute to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |

---

In [49]:
a_neutral_number = 3

In [50]:
a_neutral_number == 3 or my_favorite_number < 0

True

In [51]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [52]:
my_favorite_number < 0 or True

True

In [53]:
False or False

False

---
The `not` operator returns `False` if a condition is `True` and `True` if the condition is `False`.

---

In [54]:
not a_neutral_number == 3

False

In [55]:
not my_favorite_number < 0

True

In [56]:
not False

True

In [57]:
not True

False

Logical operators can be combined to form complex conditions. Use round brackets or parantheses `(` and `)` to indicate the order in which logical operators should be applied.

In [58]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

True

In [59]:
not (True and 0 < 1) or (False and True)

False

If parantheses are not used, logical operators are applied from left to right.

In [60]:
not True and 0 < 1 or False and True

False

---
Experiment with arithmetic, conditional and logical operators in Python using the interactive nature of Jupyter notebook. We will learn more about variables and functions in future tutorials.

---

## Adding text styles using Markdown

Adding explanations using text cells (like this one) is great way to make your notebook informative for other readers, and for yourself, if you need to refer back to it in the future. Double click on a text cell within Jupyter to edit it. In the edit mode, you'll notice that the text looks a little different (for instance the heading has a `##` prefix. This text is writted using Markdown, a simple way to add styles to your text. Execute this cell to see the output without the special characters. You can switch back and forth between the source and the output to see how to create a specific style.

For, instance, you can use one or more `#` characters at the start of a line to create headers of different sizes:

# Header 1

## Header 2

### Header 3

#### Header 4

To create a bulleted or numbered list, simply start a line with `*` or `1.`.

A bulleted list:

* Item 1
* Item 2
* Item 3

A numbered list:

1. Apple
2. Banana
3. Pineapple

You can make some text bold using `**` e.g. **this is some bold text**, or make it italic using `*` e.g. *this is some italic text.* You can also create links e.g. [this is a link](https://jovian.ml). Images are easily embedded too:

![](https://i.imgur.com/7ndD3nX.png)

Another really nice feature of Markdown is ability to include blocks of code. Note that code blocks inside Markdown cells cannot be executed.

```Python
# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

```

You can learn the full syntax of Markdown here: https://learnxinyminutes.com/docs/markdown/

## Save and upload your notebook

Whether you're running this Jupyter notebook on an online service like Binder or on your local machine, it's important to save your work from time, so that you can access it later, or share it online. You can upload this notebook to your [Jovian.ml](https://jovian.ml) account using the `jovian` Python library.

First, you need to install the Jovian python library, if it isn't already installed.

---
```Python
!pip install jovian --upgrade --quiet
```

---

Next, the library needs to be imported.

In [61]:
import jovian

Finally, you can run `jovian.commit` to capture and upload a snapshot of the notebook.

In [62]:
jovian.commit(project='W1P1-first-steps-with-python')

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Please enter your API key ( from https://jovian.ml/ ):
API KEY: 

 ·················································································································································································································································································································


[jovian] Creating a new project "uwsd/W1P1-first-steps-with-python"
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/uwsd/w1p1-first-steps-with-python


'https://jovian.ml/uwsd/w1p1-first-steps-with-python'

The first time you run `jovian.commit`, you'll be asked to provide an API Key, to securely upload the notebook to your Jovian.ml account. You can get the API key from your Jovian.ml profile page after logging in / signing up.

![api-key](https://i.imgur.com/YGGLH3e.png)

`jovian.commit` uploads the notebook to your Jovian.ml account, captures the Python environment and creates a shareable link for your notebook as shown above. You can use this link to share your work and let anyone (including you) run your notebooks and reproduce your work. Jovian also includes a powerful commenting interface, so you can discuss & comment on specific parts of your notebook:

![https://jovian.ml/docs/user-guide/upload.html](https://i.imgur.com/kxx3pqM.png)

You can do a lot more with the `jovian` Python library. Visit the documentation site to learn more: https://jovian.ml/docs/index.html

---
## Further Reading and References

Following are some resources to learn about more arithmetic, conditional and logical operations in Python:

* Python Tutorial at W3Schools: https://www.w3schools.com/python/
* Practical Python Programming: https://dabeaz-course.github.io/practical-python/Notes/Contents.html
* Python official documentation: https://docs.python.org/3/tutorial/index.html